numpy/random/mtrand/mtrand.pyx

```diff
@@ -1084,6 +1084,12 @@ cdef class RandomState:
 
         if p is not None:
             d = len(p)
+
+            atol = np.sqrt(np.finfo(np.float64).eps)
+            if isinstance(p, np.ndarray):
+                if np.issubdtype(p.dtype, np.floating):
+                    atol = max(atol, np.sqrt(np.finfo(p.dtype).eps))
+
             p = <ndarray>PyArray_ContiguousFromObject(p, NPY_DOUBLE, 1, 1)
```

Comment:
Here I think we could check if `p` can be cast to double with `NPY_SAME_KIND_CASTING`, then use `PyArray_FromAny` with `NPY_ARRAY_FORCECAST`. I wonder if Cython has a simple way to specify that?